File: Liliet/PseudoData2/TVA1_UU.cxx
#include "TVA1_UU.h"

using namespace std;  	// std namespace: so you can do things like 'cout'

ClassImp(TVA1_UU)				// classimp: necessary for root

//____________________________________________________________________________________
TVA1_UU::TVA1_UU() {
	// Default Constructor
}
//____________________________________________________________________________________
TVA1_UU::~TVA1_UU() {
	// Default Destructor
}
//____________________________________________________________________________________
Double_t TVA1_UU::TProduct(TLorentzVector v1, TLorentzVector v2) {
	// Transverse product
  Double_t tv1v2;
  return tv1v2 = v1.Px() * v2.Px() + v1.Py() * v2.Py();
}
//____________________________________________________________________________________
void TVA1_UU::SetKinematics(Double_t _QQ, Double_t _x, Double_t _t, Double_t _k){

  QQ = _QQ; //Q^2 value
  x = _x;   //Bjorken x
  t = _t;   //momentum transfer squared
  k = _k;   //Electron Beam Energy
  M2 = M*M; //Mass of the proton  squared in GeV
  //fractional energy of virtual photon
  y = QQ / ( 2. * M * k * x ); // From eq. (23) where gamma is substituted from eq (12c)
  //squared gamma variable ratio of virtuality to energy of virtual photon
	gg = 4. * M2 * x * x / QQ; // This is gamma^2 [from eq. (12c)]
  //ratio of longitudinal to transverse virtual photon flux
	e = ( 1 - y - ( y * y * (gg / 4.) ) ) / ( 1. - y + (y * y / 2.) + ( y * y * (gg / 4.) ) ); // epsilon eq. (32)
  //Skewness parameter
	xi = 1. * x * ( ( 1. + t / ( 2. * QQ ) ) / ( 2. - x + x * t / QQ ) ); // skewness parameter eq. (12b) note: there is a minus sign on the write up that shouldn't be there
  //xi = x * ( ( 1. ) / ( 2. - x ) );
  //Minimum t value
  tmin = -( QQ * ( 1. - sqrt( 1. + gg ) + gg / 2. ) ) / ( x * ( 1. - sqrt( 1. + gg ) + gg / ( 2.* x ) ) ); // minimum t eq. (29)
  //Final Lepton energy
  kpr = k * ( 1. - y ); // k' from eq. (23)
  //outgoing photon energy
  qp = t / 2. / M + k - kpr; //q' from eq. bellow to eq. (25) that has no numbering. Here nu = k - k' = k * y
  //Final proton Energy
  po = M - t / 2. / M; // This is p'_0 from eq. (28b)
  pmag = sqrt( ( -t ) * ( 1. - t / 4. / M / M ) ); // p' magnitude from eq. (28b)
  //Angular Kinematics of outgoing photon
  cth = -1. / sqrt( 1. + gg ) * ( 1. + gg / 2. * ( 1. + t / QQ ) / ( 1. + x * t / QQ ) ); // This is cos(theta) eq. (26)
	theta = acos(cth); // theta angle
  //Lepton Angle Kinematics of initial lepton
  sthl = sqrt( gg ) / sqrt( 1. + gg ) * ( sqrt ( 1. - y - y * y * gg / 4. ) ); // sin(theta_l) from eq. (22a)
	cthl = -1. / sqrt( 1. + gg ) * ( 1. + y * gg / 2. ) ; // cos(theta_l) from eq. (22a)
  //ratio of momentum transfer to proton mass
  tau = -0.25 * t / M2;

  // phi independent 4 - momenta vectors defined on eq. (21) -------------
  K.SetPxPyPzE( k * sthl, 0.0, k * cthl, k );
  KP.SetPxPyPzE( K(0), 0.0, k * ( cthl + y * sqrt( 1. + gg ) ), kpr );
  Q = K - KP;
  p.SetPxPyPzE(0.0, 0.0, 0.0, M);

  // Sets the Mandelstam variable s which is the center of mass energy
  s = (p + K) * (p + K);

  // The Gamma factor in front of the cross section
  Gamma = 1. / ALP_INV / ALP_INV / ALP_INV / PI / PI / 16. / ( s - M2 ) / ( s - M2 ) / sqrt( 1. + gg ) / x;

  // Defurne's Jacobian
  jcob = 2. * PI ;

}
//___________________________________________________________________________________
Double_t TVA1_UU::GetGamma() {

  return Gamma;
}
//___________________________________________________________________________________
void TVA1_UU::Set4VectorsPhiDep(Double_t phi) {

  // phi dependent 4 - momenta vectors defined on eq. (21) -------------

	QP.SetPxPyPzE(qp * sin(theta) * cos( phi * RAD ), qp * sin(theta) * sin( phi * RAD ), qp * cos(theta), qp);
  D = Q - QP; // delta vector eq. (12a)
  TLorentzVector pp = p + D; // p' from eq. (21)
  P = p + pp;
  P.SetPxPyPzE(.5*P.Px(), .5*P.Py(), .5*P.Pz(), .5*P.E());
}
//____________________________________________________________________________________
void TVA1_UU::Set4VectorProducts(Double_t phi) {

  // 4-vectors products (phi - independent)
  kkp  = K * KP;   //(kk')
  kq   = K * Q;    //(kq)
  kp   = K * p;    //(pk)
  kpp  = KP * p;   //(pk')

  // 4-vectors products (phi - dependent)
  kd   = K * D;    //(kΔ)
  kpd  = KP * D;   //(k'Δ)
  kP   = K * P;    //(kP)
  kpP  = KP * P;   //(k'P)
  kqp  = K * QP;   //(kq')
  kpqp = KP * QP;  //(k'q')
  dd   = D * D;    //(ΔΔ)
  Pq   = P * Q;    //(Pq)
  Pqp  = P * QP;   //(Pq')
  qd   = Q * D;    //(q
  qpd  = QP * D;   //(q'Δ)

  // Transverse vector products defined after eq.(241c) -----------------------
  kk_T   = TProduct(K,K);
  kkp_T  = kk_T;
  kqp_T  = TProduct(K,QP);
  kd_T   = -1.* kqp_T;
  dd_T   = TProduct(D,D);
  kpqp_T = kqp_T;
  kP_T   = TProduct(K,P);
  kpP_T  = TProduct(KP,P);
  qpP_T  = TProduct(QP,P);
  kpd_T  = -1.* kqp_T;
  qpd_T  = -1. * dd_T;

  //Expressions that appear in the interference coefficient calculations
  Dplus   = .5 / kpqp - .5 / kqp;
  Dminus  = -.5 / kpqp - .5 / kqp;
}
//====================================================================================
// DVCS Unpolarized Cross Section
//====================================================================================
Double_t TVA1_UU::GetDVCSUU( Double_t ReH, Double_t ReE, Double_t ReHtilde, Double_t ReEtilde, Double_t ImH, Double_t ImE, Double_t ImHtilde, Double_t ImEtilde) {

  // Note: SetKinematics should have been previously called.

  FUUT = 4.* ( ( 1 - xi * xi ) * ( ReH * ReH + ImH * ImH + ReHtilde * ReHtilde + ImHtilde * ImHtilde ) + ( tmin - t ) / ( 2.* M2 ) * ( ReE * ReE + ImE * ImE + xi * xi * ReEtilde * ReEtilde + xi * xi * ImEtilde * ImEtilde )
         - ( 2.* xi * xi ) / ( 1 - xi * xi ) * ( ReH * ReE + ImH * ImE + ReHtilde * ReEtilde + ImHtilde * ImEtilde ) );

  xdvcsUU =  GeV2nb * jcob * Gamma / QQ / ( 1 - e ) * FUUT;

  return xdvcsUU;
}
//====================================================================================
// BH Unpolarized Cross Section
//====================================================================================
Double_t TVA1_UU::GetBHUU(Double_t phi, Double_t F1, Double_t F2) {

  // Get the 4-vector products. Note: SetKinematics should have been previously called.
  Set4VectorsPhiDep(phi);
  Set4VectorProducts(phi);

  // Coefficients of the BH unpolarized structure function FUU_BH
  AUUBH = (8. * M2) / (t * kqp * kpqp) * ( (4. * tau * (kP * kP + kpP * kpP) ) - ( (tau + 1.) * (kd * kd + kpd * kpd) ) );  //eq. 147
  BUUBH = (16. * M2) / (t* kqp * kpqp) * (kd * kd + kpd * kpd); // eq. 148

  // Convert Unpolarized Coefficients to nano-barn and use Defurne's Jacobian
  con_AUUBH = AUUBH * GeV2nb * jcob;
  con_BUUBH = BUUBH * GeV2nb * jcob;

  // Unpolarized Coefficients multiplied by the Form Factors
  bhAUU = (Gamma/t) * con_AUUBH * ( F1 * F1 + tau * F2 * F2 );
  bhBUU = (Gamma/t) * con_BUUBH * ( tau * ( F1 + F2 ) * ( F1 + F2 ) );

  // Unpolarized BH cross section
  xbhUU = bhAUU + bhBUU;

  return xbhUU;
}
//====================================================================================
// Unpolarized BH-DVCS Interference Cross Section (Comparison paper)
//====================================================================================
Double_t TVA1_UU::GetIUU(Double_t phi, Double_t F1, Double_t F2, Double_t ReH, Double_t ReE, Double_t ReHtilde,
                           Double_t &AUUI, Double_t &BUUI, Double_t &CUUI) {

  // Get the 4-vector products. Note: SetKinematics should have been previously called.
  Set4VectorsPhiDep(phi);
  Set4VectorProducts(phi);

  AUUI = -4. * cos( phi * RAD ) * ( Dplus * ( ( kqp_T - 2. * kk_T - 2. * kqp ) * kpP + ( 2. * kpqp - 2. * kkp_T - kpqp_T ) * kP + kpqp * kP_T + kqp * kpP_T - 2.*kkp * kP_T ) -
                  Dminus * ( ( 2. * kkp - kpqp_T - kkp_T ) * Pqp + 2. * kkp * qpP_T - kpqp * kP_T - kqp * kpP_T ) ) ;

  BUUI = -2. * xi * cos( phi * RAD ) * ( Dplus * ( ( kqp_T - 2. * kk_T - 2. * kqp ) * kpd + ( 2. * kpqp - 2. * kkp_T - kpqp_T ) * kd + kpqp * kd_T + kqp * kpd_T - 2.*kkp * kd_T ) -
                      Dminus * ( ( 2. * kkp - kpqp_T - kkp_T ) * qpd + 2. * kkp * qpd_T - kpqp * kd_T - kqp * kpd_T ) );

  CUUI = -2. * cos( phi * RAD ) * ( Dplus * ( 2. * kkp * kd_T - kpqp * kd_T - kqp * kpd_T + 4. * xi * kkp * kP_T - 2. * xi * kpqp * kP_T - 2. * xi * kqp * kpP_T ) -
                 Dminus * ( kkp * qpd_T - kpqp * kd_T - kqp * kpd_T + 2. * xi * kkp * qpP_T - 2. * xi * kpqp * kP_T - 2. * xi * kqp * kpP_T ) );

  // Convert Unpolarized Coefficients to nano-barn and use Defurne's Jacobian
  con_AUUI = AUUI * GeV2nb * jcob;
  con_BUUI = BUUI * GeV2nb * jcob;
  con_CUUI = CUUI * GeV2nb * jcob;

  //Unpolarized Coefficients multiplied by the Form Factors
  iAUU = (Gamma/(TMath::Abs(t) * QQ)) * con_AUUI * ( F1 * ReH + tau * F2 * ReE );
  iBUU = (Gamma/(TMath::Abs(t) * QQ)) * con_BUUI * ( F1 + F2 ) * ( ReH + ReE );
  iCUU = (Gamma/(TMath::Abs(t) * QQ)) * con_CUUI * ( F1 + F2 ) * ReHtilde;

  // Unpolarized BH-DVCS interference cross section
  xIUU = iAUU + iBUU + iCUU;

  return -1. * xIUU;
}

File: Nathan/GridSearch/src/LocalFit/Parallel/TVA1_UU.go
package main

import (
	"fmt"
	"math"
	"os"
)

const ALP_INV float64 = 137.0359998 // 1 / Electromagnetic Fine Structure Constant
const PI float64 = 3.1415926535
const RAD float64 = (PI / 180.)
const M float64 = 0.938272    //Mass of the proton in GeV
const GeV2nb float64 = 389379 // Conversion from GeV to NanoBarn = .389379*1000000

type TVA1_UU struct {
	QQ, t                float64
	xi, M2, tau, e, tmin float64
	Gamma                float64
	jcob                 float64

	QP, D, P []TLorentzVector

	kkp                                               float64 // not dependent on phi
	kd, kpd, kP, kpP, kqp, kpqp, dd, Pq, Pqp, qd, qpd []float64

	kk_T, kkp_T                                                 float64 // not dependent on phi
	kqp_T, kd_T, dd_T, kpqp_T, kP_T, kpP_T, qpP_T, kpd_T, qpd_T []float64

	Dplus, Dminus []float64

	phiValues map[int]int
}

func (t *TVA1_UU) init(_QQ, x, _t, k float64, phiValuesArr []int) {
	lengthOfPhiValues := len(phiValuesArr)
	t.phiValues = make(map[int]int)

	for index, phi := range phiValuesArr {
		t.phiValues[phi] = index
	}

	t.QQ = _QQ
	t.t = _t
	t.M2 = M * M
	t.jcob = 2 * PI

	y := _QQ / (2 * M * k * x)
	gg := 4 * t.M2 * x * x / _QQ
	t.e = (1 - y - (y * y * (gg / 4))) / (1 - y + (y * y / 2) + (y * y * (gg / 4)))
	t.xi = x * (1 + _t/(2*_QQ)) / (2 - x + (x * _t / _QQ))
	t.tmin = -1 * _QQ * (1 - math.Sqrt(1+gg) + (gg / 2)) / (x * (1 - math.Sqrt(1+gg) + (gg / (2 * x))))
	kpr := k * (1 - y)
	qp := ((_t / 2) / M) + k - kpr
	cth := -1 / math.Sqrt(1+gg) * (1 + ((gg / 2) * (1 + (_t / _QQ)) / (1 + (x * _t / _QQ))))
	theta := math.Acos(cth)
	sthl := math.Sqrt(gg) / math.Sqrt(1+gg) * (math.Sqrt(1 - y - (y * y * gg / 4)))
	cthl := -1 / math.Sqrt(1+gg) * (1 + (y * gg / 2))
	t.tau = -0.25 * _t / t.M2

	K := TLorentzVector{
		k * sthl,
		0.0,
		k * cthl,
		k,
	}

	KP := TLorentzVector{
		K.parenthesesOperator(0),
		0.0,
		k * (cthl + (y * math.Sqrt(1+gg))),
		kpr,
	}

	Q := K.minus(KP)
	p := TLorentzVector{
		0.0,
		0.0,
		0.0,
		M,
	}

	s := p.plus(K).multiply(p.plus(K))
	t.Gamma = (((((((((1 / ALP_INV) / ALP_INV) / ALP_INV) / PI) / PI) / 16) / (s - t.M2)) / (s - t.M2)) / math.Sqrt(1+gg)) / x

	t.kkp = K.multiply(KP)

	t.kk_T = TVA1_UU_TProduct(K, K)
	t.kkp_T = t.kk_T

	t.QP = make([]TLorentzVector, lengthOfPhiValues)
	t.D = make([]TLorentzVector, lengthOfPhiValues)
	t.P = make([]TLorentzVector, lengthOfPhiValues)

	t.kd = make([]float64, lengthOfPhiValues)
	t.kpd = make([]float64, lengthOfPhiValues)
	t.kP = make([]float64, lengthOfPhiValues)
	t.kpP = make([]float64, lengthOfPhiValues)
	t.kqp = make([]float64, lengthOfPhiValues)
	t.kpqp = make([]float64, lengthOfPhiValues)
	t.dd = make([]float64, lengthOfPhiValues)
	t.Pq = make([]float64, lengthOfPhiValues)
	t.Pqp = make([]float64, lengthOfPhiValues)
	t.qd = make([]float64, lengthOfPhiValues)
	t.qpd = make([]float64, lengthOfPhiValues)

	t.kqp_T = make([]float64, lengthOfPhiValues)
	t.kd_T = make([]float64, lengthOfPhiValues)
	t.dd_T = make([]float64, lengthOfPhiValues)
	t.kpqp_T = make([]float64, lengthOfPhiValues)
	t.kP_T = make([]float64, lengthOfPhiValues)
	t.kpP_T = make([]float64, lengthOfPhiValues)
	t.qpP_T = make([]float64, lengthOfPhiValues)
	t.kpd_T = make([]float64, lengthOfPhiValues)
	t.qpd_T = make([]float64, lengthOfPhiValues)

	t.Dplus = make([]float64, lengthOfPhiValues)
	t.Dminus = make([]float64, lengthOfPhiValues)

	for i := 0; i < lengthOfPhiValues; i++ {
		// Set4VectorsPhiDep
		t.QP[i] = TLorentzVector{
			qp * math.Sin(theta) * math.Cos(float64(phiValuesArr[i])*RAD),
			qp * math.Sin(theta) * math.Sin(float64(phiValuesArr[i])*RAD),
			qp * math.Cos(theta),
			qp,
		}
		t.D[i] = Q.minus(t.QP[i])
		pp := p.plus(t.D[i])
		t.P[i] = p.plus(pp)
		t.P[i] = TLorentzVector{
			0.5 * t.P[i].Px(),
			0.5 * t.P[i].Py(),
			0.5 * t.P[i].Pz(),
			0.5 * t.P[i].E(),
		}

		// Set4VectorProducts
		t.kd[i] = K.multiply(t.D[i])
		t.kpd[i] = KP.multiply(t.D[i])
		t.kP[i] = K.multiply(t.P[i])
		t.kpP[i] = KP.multiply(t.P[i])
		t.kqp[i] = K.multiply(t.QP[i])
		t.kpqp[i] = KP.multiply(t.QP[i])
		t.dd[i] = t.D[i].multiply(t.D[i])
		t.Pq[i] = t.P[i].multiply(Q)
		t.Pqp[i] = t.P[i].multiply(t.QP[i])
		t.qd[i] = Q.multiply(t.D[i])
		t.qpd[i] = t.QP[i].multiply(t.D[i])

		t.kqp_T[i] = TVA1_UU_TProduct(K, t.QP[i])
		t.kd_T[i] = -1 * t.kqp_T[i]
		t.dd_T[i] = TVA1_UU_TProduct(t.D[i], t.D[i])
		t.kpqp_T[i] = t.kqp_T[i]
		t.kP_T[i] = TVA1_UU_TProduct(K, t.P[i])
		t.kpP_T[i] = TVA1_UU_TProduct(KP, t.P[i])
		t.qpP_T[i] = TVA1_UU_TProduct(t.QP[i], t.P[i])
		t.kpd_T[i] = -1 * t.kqp_T[i]
		t.qpd_T[i] = -1 * t.dd_T[i]

		t.Dplus[i] = 0.5/t.kpqp[i] - 0.5/t.kqp[i]
		t.Dminus[i] = -0.5/t.kpqp[i] - 0.5/t.kqp[i]
	}
}

func (t *TVA1_UU) getPhiIndex(phi int) int {
	index, ok := t.phiValues[phi]

	if !ok {
		fmt.Println("Error: phi value not found")
		os.Exit(1)
	}

	return index
}

func TVA1_UU_TProduct(v1, v2 TLorentzVector) float64 {
	return (v1.Px() * v2.Px()) + (v1.Py() * v2.Py())
}

func (t *TVA1_UU) getBHUU_plus_getIUU(phi int, F1, F2, ReH, ReE, ReHtilde float64) float64 {
	phiIndex := t.getPhiIndex(phi)

	// getBHUU
	AUUBH := (8. * t.M2) / (t.t * t.kqp[phiIndex] * t.kpqp[phiIndex]) * ((4. * t.tau * (t.kP[phiIndex]*t.kP[phiIndex] + t.kpP[phiIndex]*t.kpP[phiIndex])) - ((t.tau + 1.) * (t.kd[phiIndex]*t.kd[phiIndex] + t.kpd[phiIndex]*t.kpd[phiIndex])))
	BUUBH := (16. * t.M2) / (t.t * t.kqp[phiIndex] * t.kpqp[phiIndex]) * (t.kd[phiIndex]*t.kd[phiIndex] + t.kpd[phiIndex]*t.kpd[phiIndex])

	con_AUUBH := AUUBH * GeV2nb * t.jcob
	con_BUUBH := BUUBH * GeV2nb * t.jcob

	bhAUU := (t.Gamma / t.t) * con_AUUBH * (F1*F1 + t.tau*F2*F2)
	bhBUU := (t.Gamma / t.t) * con_BUUBH * (t.tau * (F1 + F2) * (F1 + F2))

	xbhUU := bhAUU + bhBUU

	// getIUU
	AUUI := -4. * math.Cos(float64(phi)*RAD) * (t.Dplus[phiIndex]*((t.kqp_T[phiIndex]-2.*t.kk_T-2.*t.kqp[phiIndex])*t.kpP[phiIndex]+(2.*t.kpqp[phiIndex]-2.*t.kkp_T-t.kpqp_T[phiIndex])*t.kP[phiIndex]+t.kpqp[phiIndex]*t.kP_T[phiIndex]+t.kqp[phiIndex]*t.kpP_T[phiIndex]-2.*t.kkp*t.kP_T[phiIndex]) -
		t.Dminus[phiIndex]*((2.*t.kkp-t.kpqp_T[phiIndex]-t.kkp_T)*t.Pqp[phiIndex]+2.*t.kkp*t.qpP_T[phiIndex]-t.kpqp[phiIndex]*t.kP_T[phiIndex]-t.kqp[phiIndex]*t.kpP_T[phiIndex]))
	BUUI := -2. * t.xi * math.Cos(float64(phi)*RAD) * (t.Dplus[phiIndex]*((t.kqp_T[phiIndex]-2.*t.kk_T-2.*t.kqp[phiIndex])*t.kpd[phiIndex]+(2.*t.kpqp[phiIndex]-2.*t.kkp_T-t.kpqp_T[phiIndex])*t.kd[phiIndex]+t.kpqp[phiIndex]*t.kd_T[phiIndex]+t.kqp[phiIndex]*t.kpd_T[phiIndex]-2.*t.kkp*t.kd_T[phiIndex]) -
		t.Dminus[phiIndex]*((2.*t.kkp-t.kpqp_T[phiIndex]-t.kkp_T)*t.qpd[phiIndex]+2.*t.kkp*t.qpd_T[phiIndex]-t.kpqp[phiIndex]*t.kd_T[phiIndex]-t.kqp[phiIndex]*t.kpd_T[phiIndex]))
	CUUI := -2. * math.Cos(float64(phi)*RAD) * (t.Dplus[phiIndex]*(2.*t.kkp*t.kd_T[phiIndex]-t.kpqp[phiIndex]*t.kd_T[phiIndex]-t.kqp[phiIndex]*t.kpd_T[phiIndex]+4.*t.xi*t.kkp*t.kP_T[phiIndex]-2.*t.xi*t.kpqp[phiIndex]*t.kP_T[phiIndex]-2.*t.xi*t.kqp[phiIndex]*t.kpP_T[phiIndex]) -
		t.Dminus[phiIndex]*(t.kkp*t.qpd_T[phiIndex]-t.kpqp[phiIndex]*t.kd_T[phiIndex]-t.kqp[phiIndex]*t.kpd_T[phiIndex]+2.*t.xi*t.kkp*t.qpP_T[phiIndex]-2.*t.xi*t.kpqp[phiIndex]*t.kP_T[phiIndex]-2.*t.xi*t.kqp[phiIndex]*t.kpP_T[phiIndex]))

	con_AUUI := AUUI * GeV2nb * t.jcob
	con_BUUI := BUUI * GeV2nb * t.jcob
	con_CUUI := CUUI * GeV2nb * t.jcob

	iAUU := (t.Gamma / (math.Abs(t.t) * t.QQ)) * con_AUUI * (F1*ReH + t.tau*F2*ReE)
	iBUU := (t.Gamma / (math.Abs(t.t) * t.QQ)) * con_BUUI * (F1 + F2) * (ReH + ReE)
	iCUU := (t.Gamma / (math.Abs(t.t) * t.QQ)) * con_CUUI * (F1 + F2) * ReHtilde

	xIUU := iAUU + iBUU + iCUU

	// getBHUU + getIUU = xbhUU + (-1 * xIUU)
	return xbhUU - xIUU
}
